### ISSUE TITLE
JSX possible issue with either no-extra-parens or indent

### ISSUE THREAD

**Issue by Moeriki** [eslint/eslint#4229]
Dears,

It seems like either one of these two rules is mistaken. I'm new to JSX so I'm not sure which one, if any.

**Attempt 1**
![image](https://cloud.githubusercontent.com/assets/38758/10672828/b85f9c08-78f2-11e5-9318-6558b6c4df1c.png)

**Attempt 2**
![image](https://cloud.githubusercontent.com/assets/38758/10672849/dbe8c6c2-78f2-11e5-839a-baacaef60d50.png)

**Attempt 3** this fixes it actually. But I liked the first two best as style.
![image](https://cloud.githubusercontent.com/assets/38758/10672912/2e74f352-78f3-11e5-8520-bee21e7676e4.png)

Your thoughts?

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
The parens issue has been brought up before (#3107) and we noted that those parens are extra, so if you want to use that style, you need to disable the rule around that area.

For indent - I think the question is whether or not `return` should imply an additional indent level or not. @gyandeeps @byk

**Comment by gyandeeps** [comment]
I am not an expert in JSX, but in attemp2 the return would only return undefined so u need braces.
As far as indent goes, indent is behaving correctly because anything below indent should align with return column wise because thats a separate statement not part of return at that time.

My answer here is based on javascript in general and not JSX.

**Comment by nzakas** [comment]
Good points. Since this has come up a few times, maybe we should just add a "jsx" option of some sort to skip these?

**Comment by btmills** [comment]
Parens around multi-line JSX seems to be the convention, and for that reason I typically disable no-extra-parens in JSX projects. I'd like to be able to re-enable it. The rule already has built-in exceptions for regular expression literals and IIFEs; would multi-line JSX have to be opt-in?

**Comment by nzakas** [comment]
Yeah, I'd expect it to be opt-in.

**Comment by btmills** [comment]
Right now no-extra-parens accepts one option: a string, either `"all"` (default), or `"functions"`. I propose adding a second option, an exceptions object. Right now it would have one possible key, `jsx`, with possible values `"never"` (default), `"all"`, or `"multiline"`. This is what it looks like added to the schema:

```js
module.exports.schema = [
    {
        "enum": ["all", "functions"]
    },
    {
        "type": "object",
        "properties": {
            "jsx": {
                "enum": ["never", "all", "multiline"]
            }
        },
        "additionalProperties": false
    }
];
```

When the `jsx` exception is unset or set to `"never"`, nothing changes, making this opt-in and non-breaking.

When the `jsx` exception is set to `"all"`, the rule will not flag excess parens around any `JSXElement`.

When the `jsx` exception is set to `"multiline"`, the rule will not flag excess parens around `JSXElement`s that span multiple lines.

```jsx
// Allowed by default/"never", "all", and "multiline"
return <foo />;

// Allowed by "all"
return (<foo />);

// Allowed by "all"
return (
    <foo />
);

// Allowed by "all" and "multiline"
return (
    <foo>
        Hello, world!
    </foo>
);
```

If this sounds like a good idea, I already have the implementation and tests ready.

**Comment by nzakas** [comment]
:+1:

**Comment by michaelficarra** [comment]
I think a better option would be `[no LineTerminator here]`, which allows parens around any multi-line expression following that spec production so it can be put on the following line.

**Comment by btmills** [comment]
@michaelficarra To confirm my understanding, you're saying that rather than being specific to JSX as proposed, the rule should be made aware of the cases where the parens are significant and stop flagging those as gratuitous?

```js
function foo() {
    return (
        someExpression
    );
}
```

Since the expression in the example is not on the same line as the return statement, the parens cannot be removed without changing the logic of the code. Right now, the rule effectively requires return values to be on the same line as the return statement. I don't know offhand where else `[no LineTerminator here]` applies, but it's doing the same thing there.

Is that right?

**Comment by michaelficarra** [comment]
@btmills Yep, that's right. Remember, it's only when an *expression* follows it, so the only productions are

    ReturnStatement[Yield]:
      return [no LineTerminator here] Expression;
      return [no LineTerminator here] Expression[In, ?Yield];

    ThrowStatement[Yield]:
      throw [no LineTerminator here] Expression[In, ?Yield];

    YieldExpression[In]:
      yield [no LineTerminator here] * AssignmentExpression[?In, Yield]
      yield [no LineTerminator here] AssignmentExpression[?In, Yield]

**Comment by btmills** [comment]
@eslint/eslint-team I think @michaelficarra makes a great point here. Right now, the rule recommends a change (deleting parentheses) that could alter the function of the code if a developer either doesn't know or doesn't remember to move the start of the contained expression to the line above. In these cases, the parentheses are significant, and I don't think it should flag them. Such a change would be non-breaking, and I might go so far to classify this as a bug we could fix. Do you all agree?

**Comment by michaelficarra** [comment]
:confused: well okay then

**Comment by nzakas** [comment]
Yeah, I think @michaelficarra's exactly right here. If we can solve this in a way that isn't specific to JSX, then let's do that.

**Comment by btmills** [comment]
I'm going to do a PR to revert #4440 until we know exactly what needs to happen here. We may not need the JSX options at all.

**Comment by satazor** [comment]
Can we have @btmills [commit](https://github.com/eslint/eslint/commit/2d32f6e381a473efd6f9272bdd25f9f325ffeae2) cherry-pick to `1.x` branch? 

I'm interested in this fix but 2.0 is not yet release :(
